### PR TITLE
Use `enum` for operation type

### PR DESCRIPTION
**Pull request by volser**

### CODE REVIEW

File: lib/quill_delta.dart [hunk 1]
```diff
@@ -30,65 +32,69 @@ class Operation {
...
+          if (operationType != OperationType.insert) return true;
           return data.length == length;
         }(), 'Length of insert operation must be equal to the text length.'),
         _attributes = attributes != null
             ? new Map<String, dynamic>.from(attributes)
             : null;
 
+  static String getOperationTypeString(OperationType opType) {
+    return opType.toString().split('.').last;
+  }
```
Comment (pulyaevskiy) [inline lib/quill_delta.dart:47]: This is fragile and the main reason I decided not to use Dart's enum in this case (I had it implemented as enum originally). What if future version of Dart changes `toString` representation of enums?

For now, I feel like enums are not the best choice when it comes to representing serializable types as they require this kind of (annoying) boilerplate.

**Review by pulyaevskiy — COMMENTED** [review]
Hi, thanks for your PR!

I'd like to understand your reasoning for wanting a enum here. Do you have a specific use case that this addresses?

I have concerns (explained in my inline comment), specifically around boilerplate code for serialization.

The way it looks now, I don't see significant value added with this change.

(On the other hand, I should've at least used constants for operation keys instead of hardcoded strings.)

File: lib/quill_delta.dart (hunk 1 above)
Comment (volser) [reply lib/quill_delta.dart:47]: the reason is to get rid of hard coded strings and allow only valid values. 
I agree, enum in Dart is crap. 

Do you want me to remade to enum class?
like:
```
class OperationType  {
  final String  value;
  toString() => value;

  const OperationType._(this.value);
  static const OperationType INSERT = const OperationType._('insert');
  static const OperationType RETAIN = const OperationType._('retain');
  static const OperationType DELETE = const OperationType._('delete');
}
```

Comment (pulyaevskiy) [reply lib/quill_delta.dart:47]: My preference would be to just simply declare constants and replace all hardcoded strings with those constants. E.g.:

```dart
class Operation {
  /// Key of insert operations.
  static const String insertKey = 'insert';

  /// Key of delete operations.
  static const String deleteKey = 'delete';

  /// Key of retain operations.
  static const String retainKey = 'retain';

  // ... rest
}
```

I feel like extra class for operation types provides little value comparing to number of lines of code added.

Would this work for you?

**Comment by volser** [comment]
I have developed quill delta in Kotlin (https://github.com/volser/android-quill-delta) and I see issue in your implementation: `insert` as string. It's not correct cuz `insert` could be any value. For example image in quill editor is `insert : { image: 'http://...'} `, so your implementation is not compatible with original format, I would like to fix that too in next PRs

File: lib/quill_delta.dart (hunk 1 above)
Comment (volser) [reply lib/quill_delta.dart:47]: obviously better than now, but not perfect )
because key is string and it not validates values

Comment (pulyaevskiy) [reply lib/quill_delta.dart:47]: The only way to create an instance of `Operation` is by using public factory constructors `Operation.insert`, `Operation.delete` and `Operation.retain` or with `Opeartion.fromJson`

All of this methods validate their inputs so there is no chance for someone to create an instance of Operation with invalid key.

Even when using an enum or a custom class the same validation would have to be implemented just in a different place. So in this case I prefer something that's simpler and less verbose.

**Comment by pulyaevskiy** [comment]
> so your implementation is not compatible with original format, I would like to fix that too in next PRs

Sure, submit a separate PR for this. Thanks!

**Comment by pulyaevskiy** [comment]
Closing this in favor of #4 .

**Pull request closed without merging**